### PR TITLE
Added metrics for failures caused by OpenStack services.

### PR DESCRIPTION
**Pull request by gryf**
In Kuryr-kubernetes we mainly rely on two services: Neutron for ports
creating and Octavia for load balancers. Sometimes we observe that either
ports or load balancer hangs in some states indefinitely. In such case,
after we timeout, kuryr controller will crash, leaving ambiguous log
message.

In this patch we introduce two new Prometheus metrics - one for Port and
the other for load balancer, which will be updated during such
situation.

### CODE REVIEW

File: kuryr_kubernetes/controller/managers/prometheus_exporter.py [hunk 1]
```diff
@@ -134,3 +142,13 @@ def _create_metrics(self):
         self.pod_creation_latency = prometheus_client.Histogram(
             'kuryr_pod_creation_latency', 'Time taken for a pod to have'
             ' Kuryr annotations set', buckets=buckets, registry=self.registry)
+
+        self.load_balancer_readyness = prometheus_client.Enum(
+            'kuryr_load_balancer_readyness',
+            'Indicates issues with creating Octavia LB',
+            states=['ok', 'notready'], registry=self.registry)
```
Comment (MaysaMacedo) [inline kuryr_kubernetes/controller/managers/prometheus_exporter.py:149]: Should the 'ok' stauts be ever set? or perhaps the metric type should be a Gauge?

**Review by MaysaMacedo — COMMENTED** [review]
Looks good. Just a couple of suggestions:

File: kuryr_kubernetes/controller/managers/prometheus_exporter.py (hunk 1 above)
Comment (dulek) [reply kuryr_kubernetes/controller/managers/prometheus_exporter.py:149]: +1, we need a strategy to get the problematic status cleared up.

File: kuryr_kubernetes/controller/managers/prometheus_exporter.py [hunk 2]
```diff
@@ -134,3 +142,13 @@ def _create_metrics(self):
         self.pod_creation_latency = prometheus_client.Histogram(
             'kuryr_pod_creation_latency', 'Time taken for a pod to have'
             ' Kuryr annotations set', buckets=buckets, registry=self.registry)
+
+        self.load_balancer_readyness = prometheus_client.Enum(
```
[the remark this answers is not in the document]
Comment (gryf) [reply kuryr_kubernetes/controller/managers/prometheus_exporter.py:146]: If we create labeled Enum for every each created LB, there would be a ton of them. I guess that it's not what we want. All the details regarding failed LB/port are in logs, therefore I'd keep it that way.

File: kuryr_kubernetes/controller/managers/prometheus_exporter.py (hunk 1 above)
Comment (gryf) [reply kuryr_kubernetes/controller/managers/prometheus_exporter.py:149]: During creation of the Enum type, the state is automatically set to the first one from the list.
As for status reset - after the kuryr controller restart it should "reset" itself. Unless we have some other way for cleaning failed port/lb, I can't see any other option for removing an alert.

File: kuryr_kubernetes/controller/managers/prometheus_exporter.py [hunk 3]
```diff
@@ -134,3 +142,13 @@ def _create_metrics(self):
...
+            'kuryr_load_balancer_readyness',
+            'Indicates issues with creating Octavia LB',
+            states=['ok', 'notready'], registry=self.registry)
+
+        self.port_readyness = prometheus_client.Enum(
```
[the remark this answers is not in the document]
Comment (gryf) [reply kuryr_kubernetes/controller/managers/prometheus_exporter.py:151]: argh :)

File: kuryr_kubernetes/controller/managers/prometheus_exporter.py [hunk 4]
```diff
@@ -134,3 +142,13 @@ def _create_metrics(self):
...
+            'kuryr_port_readyness',
+            'Indicates issues with creating port in Neutron',
```
[the remark this answers is not in the document]
Comment (gryf) [reply kuryr_kubernetes/controller/managers/prometheus_exporter.py:153]: Response for that comment also (or even more, we create much more ports ;) applies here.

File: kuryr_kubernetes/controller/managers/prometheus_exporter.py [hunk 5]
```diff
@@ -134,3 +142,12 @@ def _create_metrics(self):
         self.pod_creation_latency = prometheus_client.Histogram(
             'kuryr_pod_creation_latency', 'Time taken for a pod to have'
             ' Kuryr annotations set', buckets=buckets, registry=self.registry)
+
+        self.load_balancer_readiness = prometheus_client.Count(
+            'kuryr_load_balancer_readiness',
+            'Increasing this metric indicates issues with creating Octavia LB',
```
Comment (dulek) [inline kuryr_kubernetes/controller/managers/prometheus_exporter.py:148]: How about : "This metric is increased when Kuryr notices that an Octavia load balancer is stuck in an unexpected state"?

File: kuryr_kubernetes/controller/managers/prometheus_exporter.py [hunk 6]
```diff
@@ -134,3 +142,12 @@ def _create_metrics(self):
...
+            registry=self.registry)
+
+        self.port_readiness = prometheus_client.Count(
+            'kuryr_port_readiness', 'Increasing this metric indicates issues '
```
Comment (dulek) [inline kuryr_kubernetes/controller/managers/prometheus_exporter.py:152]: Maybe "This counter is increased when Kuryr times out waiting for Neutron to move port to ACTIVE"?

**Comment by dulek** [comment]
Looks good, just minor remarks on wording. I'm not 100% confident mine's better though.

**Comment by MaysaMacedo** [comment]
Looks good and ocp installation passed, but destroy failed, probably due to the octavia issue.

**Pull request merged**